### PR TITLE
V3.5.1 zqb skinning

### PR DESCRIPTION
**Pull request by zhakesi**
1] 根据硬件查询maxVertexUniformVectors 设置最大joints数量
2] maxVertexUniformVectors 不足时使用texture代替UniformVectors
为避免传输数据过大(例如PC可能最大支持4096VertexUniformVectors) 最大joints数量限定在256

*  https://github.com/cocos/3d-tasks/issues/5558

### CODE REVIEW

File: native/cocos/renderer/pipeline/Define.h [hunk 1]
```diff
@@ -330,14 +333,15 @@ struct CC_DLL UBOSkinningAnimation {
...
-    static constexpr uint                        SIZE          = UBOSkinning::COUNT * 4;
-    static constexpr uint                        BINDING       = static_cast<uint>(ModelLocalBindings::UBO_SKINNING_TEXTURE);
+class CC_DLL UBOSkinning {
+public:
+    static uint                                  count;
+    static uint                                  size;
+    static constexpr uint                        BINDING            = static_cast<uint>(ModelLocalBindings::UBO_SKINNING_TEXTURE);
     static const gfx::DescriptorSetLayoutBinding DESCRIPTOR;
-    static const gfx::UniformBlock               LAYOUT;
+    static gfx::UniformBlock                     layout;
```
Comment (star-e) [inline native/cocos/renderer/pipeline/Define.h:342]: Does `gfx::UniformBlock layout` change over time? It is not `const` any more.

Comment (zhakesi) [reply native/cocos/renderer/pipeline/Define.h:342]: > Does `gfx::UniformBlock layout` change over time? It is not `const` any more.

 do you mean just changing LAYOUT‘s member variable is a better way?

Comment (star-e) [reply native/cocos/renderer/pipeline/Define.h:342]: After engine initialized, if user called some related API, will the content change in `layout`?

Comment (zhakesi) [reply native/cocos/renderer/pipeline/Define.h:342]: yes. now in engine it's like this. or i should make sure InitLayout just be called once.
```
    public static InitLayout (capacity: number) {
        UBOSkinning._JOINT_UNIFORM_CAPACITY = capacity;
        UBOSkinning._COUNT = capacity * 12;
        UBOSkinning._SIZE = UBOSkinning._COUNT * 4;
        UBOSkinning._LAYOUT = new UniformBlock(SetIndex.LOCAL, UBOSkinning.BINDING, UBOSkinning.NAME, [
            new Uniform('cc_joints', Type.FLOAT4, capacity * 3),
        ], 1);
    }

/**
 * @internal This method only used in Game._initDevice to init localDescriptorSetLayout.layouts[UBOSkinning.NAME]
*/
export function localDescriptorSetLayout_ResizeMaxJoints (maxCount: number) {
    UBOSkinning.InitLayout(maxCount);
    localDescriptorSetLayout.layouts[UBOSkinning.NAME] = UBOSkinning.LAYOUT;
    localDescriptorSetLayout.bindings[UBOSkinning.BINDING] = UBOSkinning.DESCRIPTOR;
}
```

File: native/cocos/renderer/GFXDeviceManager.h [hunk 2]
```diff
@@ -64,7 +64,7 @@ namespace gfx {
 
 class CC_DLL DeviceManager final {
     static constexpr bool DETACH_DEVICE_THREAD{true};
-    static constexpr bool FORCE_DISABLE_VALIDATION{false};
+    static constexpr bool FORCE_DISABLE_VALIDATION{true};
```
Comment (dumganhar) [inline native/cocos/renderer/GFXDeviceManager.h]: Why set this flag to true?

Comment (zhakesi) [reply native/cocos/renderer/GFXDeviceManager.h]: @timlyeee

Comment (zhakesi) [reply native/cocos/renderer/GFXDeviceManager.h]: > Why set this flag to true?

这部分是李欣提交的commit. 
他是说这样可以跳过gfx对象校验可以降低tick运行损耗

Comment (pandamicro) [reply native/cocos/renderer/GFXDeviceManager.h]: Shouldn't submit this change in the current PR, focus on the problem you are solving

File: cocos/core/assets/effect-asset.ts [hunk 3]
```diff
@@ -294,6 +295,14 @@ export class EffectAsset extends Asset {
     public validate () {
         return this.techniques.length > 0 && this.shaders.length > 0;
     }
+
+    private modifyDescriptorSetLayout () {
```
Comment (pandamicro) [inline cocos/core/assets/effect-asset.ts:299]: use `_` as starter for private function names

File: cocos/core/assets/effect-asset.ts [hunk 4]
```diff
@@ -294,6 +295,14 @@ export class EffectAsset extends Asset {
...
+        //set max joints after device initialize.
+        const root = legacyCC.director.root as Root;
+        let maxJoints = Math.floor((root.device.capabilities.maxVertexUniformVectors - 38) / 3);
+        maxJoints = maxJoints < 256 ? maxJoints : 256;
+        localDescriptorSetLayout_ResizeMaxJoints(maxJoints);
```
Comment (pandamicro) [inline cocos/core/assets/effect-asset.ts:304]: This is too hack, in general EffectAsset, we shouldn't care about the joint layout or max joints availability, we should move the check to other places

File: native/cocos/renderer/pipeline/GlobalDescriptorSetManager.cpp [hunk 5]
```diff
@@ -61,6 +61,11 @@ void GlobalDSManager::activate(gfx::Device *device, RenderPipeline *pipeline) {
         gfx::Address::CLAMP,
     });
 
+    uint maxJoints = (_device->getCapabilities().maxVertexUniformVectors - 38) / 3;
+    maxJoints = maxJoints < 256 ? maxJoints : 256;
+    SkinningJointCapacity::jointUniformCapacity = maxJoints;
+    UBOSkinning::initLayout(maxJoints);
```
Comment (pandamicro) [inline native/cocos/renderer/pipeline/GlobalDescriptorSetManager.cpp:67]: I think it's ok the initialized UBOSkinning layout here, please use the same approach in TypeScript engine

**Review by pandamicro — CHANGES_REQUESTED** [review]
You shouldn't delete any unrelated chunks or effects in surfaces. And auto bindings shouldn't be submitted either

**Pull request closed without merging**